File: server/bin/main/com/example/reconix/server/service/PurchaseOrderService.kt
package com.example.reconix.server.service

import com.example.reconix.server.database.*
import com.example.reconix.shared.*
import kotlinx.datetime.Clock
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.transactions.transaction

/**
 * Purchase Order Service - Handles PO creation, PDF generation, and email sending
 */
class PurchaseOrderService {

    private val pdfService = PdfGeneratorService()
    private val emailService = EmailService()

    /**
     * Create a new Purchase Order
     * 1. Insert PO + items into database
     * 2. Generate PO PDF
     * 3. Email PDF to vendor (if configured)
     * 4. Return response
     */
    fun createPurchaseOrder(request: CreatePurchaseOrderRequest): CreatePurchaseOrderResponse {
        return transaction {
            val now = Clock.System.now().toString()

            // Generate PO ID: PO-XXX (sequential)
            val lastPo = PurchaseOrders.selectAll()
                .orderBy(PurchaseOrders.id, SortOrder.DESC)
                .limit(1)
                .map { it[PurchaseOrders.id] }
                .firstOrNull()

            val nextNum = if (lastPo != null) {
                val num = lastPo.removePrefix("PO-").toIntOrNull() ?: 0
                num + 1
            } else {
                1
            }
            val poId = "PO-${String.format("%03d", nextNum)}"

            // Calculate total amount
            val totalAmount = request.items.sumOf { item ->
                val subtotal = item.quantity * item.unitPrice
                subtotal + (subtotal * item.taxRate / 100.0)
            }

            // Insert Purchase Order
            PurchaseOrders.insert {
                it[id] = poId
                it[vendorName] = request.vendorName
                it[vendorEmail] = request.vendorEmail.ifBlank { null }
                it[PurchaseOrders.totalAmount] = totalAmount
                it[status] = "OPEN"
                it[createdAt] = now
            }

            // Insert line items
            request.items.forEachIndexed { index, item ->
                val itemId = "ITEM-${poId}-${String.format("%02d", index + 1)}"
                PurchaseOrderItems.insert {
                    it[PurchaseOrderItems.poId] = poId
                    it[PurchaseOrderItems.itemId] = itemId
                    it[itemName] = item.itemName
                    it[quantity] = item.quantity
                    it[unitPrice] = item.unitPrice
                }
            }

            // Generate PDF
            var pdfGenerated = false
            var pdfPath: String? = null
            try {
                pdfPath = pdfService.generatePoPdf(
                    poId = poId,
                    vendorName = request.vendorName,
                    vendorEmail = request.vendorEmail,
                    items = request.items.map { PdfLineItem(it.itemName, it.quantity, it.unitPrice, it.taxRate) },
                    totalAmount = totalAmount
                )
                pdfGenerated = true

                // Update PO with PDF path
                PurchaseOrders.update({ PurchaseOrders.id eq poId }) {
                    it[PurchaseOrders.pdfPath] = pdfPath
                }
            } catch (e: Exception) {
                println("⚠️ PDF generation failed: ${e.message}")
            }

            // Send email to vendor
            var emailSent = false
            if (request.vendorEmail.isNotBlank() && pdfPath != null) {
                emailSent = emailService.sendPoToVendor(request.vendorEmail, poId, pdfPath)
                if (emailSent) {
                    PurchaseOrders.update({ PurchaseOrders.id eq poId }) {
                        it[status] = "SENT"
                    }
                }
            }

            CreatePurchaseOrderResponse(
                success = true,
                poId = poId,
                status = if (emailSent) "SENT" else "OPEN",
                message = buildString {
                    append("Purchase Order $poId created successfully")
                    if (pdfGenerated) append(". PDF generated")
                    if (emailSent) append(". Email sent to ${request.vendorEmail}")
                    else if (request.vendorEmail.isNotBlank()) append(". Email not sent (SMTP not configured)")
                },
                totalAmount = totalAmount,
                pdfGenerated = pdfGenerated,
                emailSent = emailSent
            )
        }
    }
}

File: python_server/services/purchase_order_service.py
"""
services/purchase_order_service.py - Create POs, generate PDF, send email
Equivalent to server/.../service/PurchaseOrderService.kt
"""
from datetime import datetime, timezone
from typing import List

from sqlalchemy.orm import Session

from models import PurchaseOrder, PurchaseOrderItem
from schemas import CreatePOLineItem, CreatePurchaseOrderRequest, CreatePurchaseOrderResponse
from services.pdf_generator_service import PdfLineItem, generate_po_pdf
from services.email_service import send_po_to_vendor


def create_purchase_order(db: Session, request: CreatePurchaseOrderRequest) -> CreatePurchaseOrderResponse:
    now = datetime.now(timezone.utc).isoformat()

    # Determine next PO number
    last_po = db.query(PurchaseOrder).order_by(PurchaseOrder.id.desc()).first()
    if last_po:
        try:
            next_num = int(last_po.id.replace("PO-", "")) + 1
        except ValueError:
            next_num = 1
    else:
        next_num = 1
    po_id = f"PO-{next_num:03d}"

    # Calculate total (including tax)
    total_amount = sum(
        item.quantity * item.unitPrice * (1 + item.taxRate / 100.0)
        for item in request.items
    )

    # Insert PO
    po = PurchaseOrder(
        id=po_id,
        vendor_name=request.vendorName,
        vendor_email=request.vendorEmail or None,
        total_amount=total_amount,
        status="OPEN",
        created_at=now,
    )
    db.add(po)
    db.flush()

    # Insert PO items
    for idx, item in enumerate(request.items):
        item_id = f"ITEM-{po_id}-{idx + 1:02d}"
        db.add(PurchaseOrderItem(
            po_id=po_id,
            item_id=item_id,
            item_name=item.itemName,
            quantity=item.quantity,
            unit_price=item.unitPrice,
        ))
    db.flush()

    # Generate PDF
    pdf_generated = False
    pdf_path: str | None = None
    try:
        pdf_line_items = [
            PdfLineItem(
                item_name=item.itemName,
                quantity=item.quantity,
                unit_price=item.unitPrice,
                tax_rate=item.taxRate,
            )
            for item in request.items
        ]
        pdf_path = generate_po_pdf(
            po_id=po_id,
            vendor_name=request.vendorName,
            vendor_email=request.vendorEmail,
            items=pdf_line_items,
            total_amount=total_amount,
        )
        pdf_generated = True
        po.pdf_path = pdf_path
    except Exception as exc:
        print(f"⚠️ PDF generation failed: {exc}")

    # Send email
    email_sent = False
    if request.vendorEmail and pdf_path:
        email_sent = send_po_to_vendor(
            vendor_email=request.vendorEmail,
            po_id=po_id,
            pdf_path=pdf_path,
        )
        if email_sent:
            po.status = "SENT"

    db.commit()

    # Build response message
    parts = [f"Purchase Order {po_id} created successfully"]
    if pdf_generated:
        parts.append("PDF generated")
    if email_sent:
        parts.append(f"Email sent to {request.vendorEmail}")
    elif request.vendorEmail:
        parts.append("Email not sent (SMTP not configured)")

    return CreatePurchaseOrderResponse(
        success=True,
        poId=po_id,
        status="SENT" if email_sent else "OPEN",
        message=". ".join(parts),
        totalAmount=total_amount,
        pdfGenerated=pdf_generated,
        emailSent=email_sent,
    )
